lib/byron/test/unit/Cardano/Wallet/Byron/TransactionSpec.hs

```diff
@@ -159,6 +223,197 @@ spec = do
+
+        -- We always go for the higher bound for change address payload's size,
+        -- so, we may end up with up to 12 extra bytes per change address in our
+        -- estimation.
+        -- For Icarus addresses, we can be as good as 4 bytes per change address
+        -- because there's no variance due to the derivation path encoded as
+        -- attributes (this only happens on random addresses).
+        margin = 12 * fromIntegral (length $ change sel)
+        realSizeSup = Quantity (size + margin)
+        realSizeInf = Quantity size
+    in
+        property (calcSize >= realSizeInf && calcSize <= realSizeSup)
+            & counterexample ("Raw Transaction:" <> CBOR.prettyHexEnc cbor)
+            & counterexample ("Coin Selection:\n" <> pretty sel)
+            & counterexample (unlines
+                [ "real size inf:   " <> show size
+                , "real size sup:   " <> show (size + margin)
+                , "estimated size:  " <> show (getQuantity calcSize)
+                ])
+  where
+    fromCoinSelection :: CoinSelection -> [Address] -> CBOR.Encoding
+    fromCoinSelection (CoinSelection inps outs chngs) chngAddrs =
+        CBOR.encodeSignedTx (fst <$> inps, outs <> outs') wits
+      where
+        dummySig =
+            BS.replicate 64 0
+        Right dummyXPub =
+            xpub $ BS.replicate 64 0
+        outs' =
+            zipWith TxOut (take (length chngs) chngAddrs) chngs
+        wits =
+            replicate (length inps)
+                $ CBOR.toStrictByteString
+                $ CBOR.encodePublicKeyWitness dummyXPub dummySig
+
+newtype LegacyAddress (n :: NetworkDiscriminant) (k :: Depth -> * -> *) =
+    LegacyAddress (Proxy n, Proxy k, Address)
+
+instance (PaymentAddress n ByronKey) => Arbitrary (LegacyAddress n ByronKey)
+  where
+    arbitrary = do
+        bytes <- BS.pack <$> vector 64
+        addrIx <- toEnum . fromIntegral <$> choose (0, 10 :: Word32)
+        let (Right key) = xpub bytes
+        let pw = Passphrase $ BA.convert $ BS.replicate 32 0
+        pure $ LegacyAddress
+            ( Proxy
+            , Proxy
+            , paymentAddress @n $ ByronKey key (minBound, addrIx) pw
+            )
+
+instance (PaymentAddress n IcarusKey) => Arbitrary (LegacyAddress n IcarusKey)
+  where
+    arbitrary = do
+        bytes <- BS.pack <$> vector 64
+        let (Right key) = xpub bytes
+        pure $ LegacyAddress
+            ( Proxy
+            , Proxy
+            , paymentAddress @n $ IcarusKey key
+            )
+
+genLegacyAddress
+    :: forall (n :: NetworkDiscriminant) k.
+        ( PaymentAddress n k
+        , Arbitrary (LegacyAddress n k)
+        )
+    => Gen Address
+genLegacyAddress = do
+    LegacyAddress (_ :: Proxy n, _ :: Proxy k, addr) <- arbitrary
+    pure addr
+
+-- 'CoinSelection' are generated by running the @largestFirst@ algorithm on an
+-- arbitrary UTxO and to cover an arbitrary set of 'TxOut'.
+genSelection
```

Comment:
Nice generators. I was wondering if it is actually necessary to generate a UTxO then run a coin selection, or are CoinSelection values enough.

Reply:
I tried generating CoinSelection at first, but there are just so many cases for invalid coin selections... Since it's quite easy however to generate valid UTxOs, I went down this path.